0015-authority.md

```diff
@@ -0,0 +1,478 @@
+        NetworkManager.MessageSender.Send(NetworkConstants.CHANGE_AUTHORITY, NetworkChannel.Internal, buffer);
+    }
+}
+```
+
+Wrapper function for NetworkObject change in NetworkSpawnManager:
+```csharp
+public void ChangeAuthority(Authority authority)
+{
+    if (networkObject.IsSpawned == false)
+    {
+        throw new SpawnStateException("Object is not spawned");
+    }
+
+    // convert flags to index
+    var serialized = new BitField32((uint)authority).CountLeadingZeros();
+    ChangeAuthority(networkObject.NetworkObjectId, serialized);
+}
+```
+
+The deserialization side will contain similar logic as change ownership and will apply authority changes to the NetworkObject.
+
+## NetworkVariable Authority
+
+We introduce a `NetworkVariablePermission.Authority` permission. This is a very simple change to the NetworkVariable we just add the following to the read/write queries
+```csharp
+public bool CanClientRead(ulong clientId)
+{
+    switch (Settings.ReadPermission)
+    {
+    //....
+        case NetworkVariablePermission.Authority:
+            return m_NetworkBehaviour.HasAuthority;
+    //.....
+    }
+    return true;
+}
+```
+
+## OnGainedAuthority / OnLostAuthority
+
+Will be implemented by listening to authority mode changes internally and then checking whether the resulting `HasAuthority` flag is different.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- The authority system is very loosely integrated into MLAPI. It does not change how NetworkVariables and RPCs work etc. While there are many upsides to keeping authority as a separate abstraction layer it could also be a missed opportunity depending on the direction we want to take MLAPI.
+
+- An ACL system could potentially change how we think about ownership or authority or at least allow us to implement the authority system with much less code changes needed. Waiting for the ACL system could be a good idea.
+
+- Authority increases the spawn header size of a `NetworkObject` by 2 bit and the header size of `NetworkBehaviours` which use the `AllowAuthorityOverride` by 2 bit. In the future once we have more authority modes these headers will potentially get even larger.
+
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+- Why is this design the best in the space of possible designs?
+- What other designs have been considered and what is the rationale for not choosing them?
+- What is the impact of not doing this?
+
+Authority could have been done as a permission/acl system instead of an authority mode. This would have allowed us to have more fine-grained control over who has authority. We were not able to find many real world use cases where this would be useful so that's why the authority mode was chosen. In the future we might consider storing authority internally as a set of owners instead of an authority mode but then most likely we would still want to expose authority as an authority mode for ease of use.
```

Comment:
Well, I think ACL wouldn't even need to be exposed to the end-user directly. It could be a backbone that we use internally and extend over time with more props, flags, queries etc.
In the absence of a full-fledged and generalized ACL concept within MLAPI internals, I believe we could address these usage-patterns with ad-hoc implementations as @SamuelBellomo did in `NetworkTransform`.
The reason why I'm saying that is basically the fact that you also mentioned this Authority Model wouldn't fit in all places and we'll probably keep Ownership-over-RPC and most parts of NetVar-ReadWrite-Perms for a while.
I'm questioning whether we add more things to replace later here, or in other terms, I can't really see this model being durable for a long tme.